Feat(eos_cli_config_gen): Add support for mpls tunnel termination settings

ccsnw opened the pull request:
## Change Summary

Add knobs to configure tunnel termination settings for MPLS

## Related Issue(s)

Fixes #<N/A>

## Component(s) name

`arista.avd.eos_cli_config_gen`

## Proposed changes
      tunnel:
        description: Configure tunnel termination model.
        type: dict
        keys:
          termination_model:
            type: str
            description: |
              TTL and DSCP configuration as a string.
              Example:
              - "ttl pipe dscp pipe"
            valid_values:
              - "ttl pipe dscp pipe"
              - "ttl uniform dscp pipe"
              - "ttl uniform dscp uniform"
          termination_php_model:
            type: str
            description: |
              TTL and DSCP configuration as a string.
              Example:
              - "ttl pipe dscp pipe"
            valid_values:
              - "ttl pipe dscp pipe"
              - "ttl uniform dscp pipe"

## How to test
added molecule test
cli command order was tested on EOS 4.30.5 - **Please let me know if newer version sort differently**

## Checklist

### User Checklist

<!-- Add your own checklist using MD syntax and by replacing N/A -->
- N/A

### Repository Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code has been rebased from devel before I start
- [x] I have read the [**CONTRIBUTING**](https://avd.sh/en/latest/docs/contribution/overview.html) document.
- [x] My change requires a change to the documentation and documentation have been updated accordingly.
- [x] I have updated [molecule CI](https://github.com/aristanetworks/avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly. (check the box if not applicable)

ccsnw committed:
add_mpls_tunnel_termination

MaheshGSLAB commented on python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/mpls.schema.yml:
```diff
@@ -282,3 +282,26 @@ keys:
+              TTL and DSCP configuration as a string.
+              Example:
+              - "ttl pipe dscp pipe"
+            valid_values:
+              - "ttl pipe dscp pipe"
+              - "ttl uniform dscp pipe"
```
```suggestion
            valid_values:
              - "ttl pipe dscp pipe"
              - "ttl pipe dscp uniform"
              - "ttl uniform dscp pipe"
              - "ttl uniform dscp uniform"
```
```
wan1-dc1(config)#mpls tunnel termination php model ttl pipe dscp ?
  pipe     Preserve the inner value
  uniform  Propagate value from outer header

wan1-dc1(config)#mpls tunnel termination php model ttl uniform ?
  dscp  model for dscp

wan1-dc1(config)#mpls tunnel termination php model ttl uniform dscp ?
  pipe     Preserve the inner value
  uniform  Propagate value from outer header

```

ccsnw replied:
@MaheshGSLAB which EOS did you try this on? 

I've tested on 4.31.4M (Platform 7289-F) & 4.26.3M (DCS-7280CR2A-30-F ).

The results are different e.g.:

4.31.4M (Platform 7289-F):

```
X(s1)(config-s-test)#mpls tunnel termination php model ttl ?
  pipe     Preserve the inner value
  uniform  Propagate value from outer header

X(s1)(config-s-test)#mpls tunnel termination php model ttl pipe dscp ?
  pipe  Preserve the inner value

X(s1)(config-s-test)#mpls tunnel termination php model ttl uniform dscp ?
  pipe  Preserve the inner value
```

Is there a way to find out if the commandset depends on Platform or EOS-Version?

I think we can list all options that you've proposed for both models, so

```
              - "ttl pipe dscp pipe"
              - "ttl pipe dscp uniform"
              - "ttl uniform dscp pipe"
              - "ttl uniform dscp uniform"
```

and I can add an info to the description that says: Depends on Platform/EOS ??

ccsnw replied:
added all 4 choices for both models and a hint in the description to check for platform compatibility. Please reopen if you have a better idea on how to handle this or are not Ok with the solution.

MaheshGSLAB replied:
Hi @ccsnw 
Am using cEOS 4.33.2F.

MaheshGSLAB replied:
Also you can run show cli commands to get number of supported commands
```
[no|default] mpls tunnel termination
[no|default] mpls tunnel termination model ttl TTL_MODEL dscp DSCP_MODEL
[no|default] mpls tunnel termination php model ttl TTL_MODEL dscp DSCP_MODEL
```

ccsnw replied:
thanks, wasn't aware of that command. its useful!

MaheshGSLAB commented on ansible_collections/arista/avd/molecule/eos_cli_config_gen/inventory/host_vars/host1/mpls.yml:
```diff
@@ -6,6 +6,8 @@ mpls:
     shutdown: false
     transport_address_interface: Loopback0
+  tunnel:
+    termination_model: "ttl uniform dscp uniform"
```
```suggestion
    termination_model: ttl uniform dscp uniform
    termination_php_model: ttl pipe dscp pipe
```

ccsnw committed:
add_mpls_tunnel_termination

ClausHolbechArista reviewed (changes requested):
I would like a more elaborate data model instead of all these valid values in a multi-word string. I do not know the CLI here in detail, so @ccsnw if you could provide a suggestion, we can take a look. Otherwise we can investigate more and get back.

ccsnw commented:
> I would like a more elaborate data model instead of all these valid values in a multi-word string. I do not know the CLI here in detail, so @ccsnw if you could provide a suggestion, we can take a look. Otherwise we can investigate more and get back.

Hi @ClausHolbechArista 

Not an expert on that CLI either. Colleagues are needing it on some WAN routers.

But I noticed that the modes (php vs no php) are not mutually exclusive, so we can have sth like:

```
mpls tunnel termination model ttl pipe dscp pipe
mpls tunnel termination php model ttl pipe dscp pipe
```

in the same config.

Hence I'd suggest a model like this:

```
tunnel:
  termination:
    - php: < true | false >
      model:   
        ttl: < pipe | uniform >
        dscp: < pipe | uniform >
```

We can leave it up to the user to decide with model config makes sense (and actually is supported).

What do you think?

alexeygorbunov commented:
> ```
> tunnel:
>   termination:
>     - php: < true | false >
>       model:   
>         ttl: < pipe | uniform >
>         dscp: < pipe | uniform >
> ```

Or something explicit like this:

```yaml
      tunnel:
        description: Configure MPLS tunnel.
        type: dict
        keys:
          termination:
            description: Controls selection of the TTL/DSCP values by LER when decapsulating MPLS packets.
            type: dict
            keys:
              model:
                type: dict
                keys:
                  ttl:
                    type: str
                    required: true
                    valid_values:
                      - pipe
                      - uniform
                  dscp:
                    type: str
                    required: true
                    valid_values:
                      - pipe
              php_model:
                description: Used on PHP router in the absence of any VPN routes and explicit null VRF labels.
                type: dict
                keys:
                  ttl:
                    type: str
                    required: true
                    valid_values:
                      - pipe
                      - uniform
                  dscp:
                    type: str
                    required: true
                    valid_values:
                      - pipe
                      - uniform
```

I was not able to identify any specific platforms which support `dscp uniform` when configuring `mpls tunnel termination model ttl`. I'll try to figure this out . If not - we may need to only permit `pipe` under `mpls.tunnel.termination.model.dscp`.

Update: I got a confirmation that `dscp uniform` mode is not supported in non-PHP use case.

ccsnw commented:
> > ```
> > tunnel:
> >   termination:
> >     - php: < true | false >
> >       model:   
> >         ttl: < pipe | uniform >
> >         dscp: < pipe | uniform >
> > ```
> 
> Or something explicit like this:
> 
> ```yaml
>       tunnel:
>         description: Configure MPLS tunnel.
>         type: dict
>         keys:
>           termination:
>             description: Controls selection of the TTL/DSCP values by LER when decapsulating MPLS packets.
>             type: dict
>             keys:
>               model:
>                 type: dict
>                 keys:
>                   ttl:
>                     type: str
>                     required: true
>                     valid_values:
>                       - pipe
>                       - uniform
>                   dscp:
>                     type: str
>                     required: true
>                     valid_values:
>                       - pipe
>                       - uniform # support needs to be confirmed
>               php_model:
>                 description: Used on PHP router in the absence of any VPN routes and explicit null VRF labels.
>                 type: dict
>                 keys:
>                   ttl:
>                     type: str
>                     required: true
>                     valid_values:
>                       - pipe
>                       - uniform
>                   dscp:
>                     type: str
>                     required: true
>                     valid_values:
>                       - pipe
>                       - uniform
> ```
> 
> I was not able to identify any specific platforms which support `dscp uniform` when configuring `mpls tunnel termination model ttl`. I'll try to figure this out . If not - we may need to only permit `pipe` under `mpls.tunnel.termination.model.dscp`.

Hi Alexey,

The explicit version also looks fine to me.

Regarding compatibility. Check this:

```
X(config-s-test)#mpls tunnel termination model ttl pipe dscp ?
  pipe     Preserve the inner value
  uniform  Propagate value from outer header

X(config-s-test)#mpls tunnel termination model ttl pipe dscp uniform
% ttl model pipe is not supported with dscp model uniform

X(config-s-test)#sh version
Arista DCS-7280CR2A-30-F
Hardware version: 21.00
<snip>
Software image version: 4.26.3M
```

On 4.26.3 the "dscp uniform" was suggested by eos, but couldn't be configured (not supported)

When i check on a newer EOS:

```
X(s1)(config-s-test)#mpls tunnel termination model ttl pipe dscp ?
  pipe  Preserve the inner value

X(s1)(config-s-test)#sh ver
Arista 7289-F
<snip>
Software image version: 4.31.4M
```

not suggesting dscp > uniform anymore.

MaheshGSLAB commented:
> ```yaml
> tunnel:
>         description: Configure MPLS tunnel.
>         type: dict
>         keys:
>           termination:
>             description: Controls selection of the TTL/DSCP values by LER when decapsulating MPLS packets.
>             type: dict
>             keys:
>               model:
>                 type: dict
>                 keys:
>                   ttl:
>                     type: str
>                     required: true
>                     valid_values:
>                       - pipe
>                       - uniform
>                   dscp:
>                     type: str
>                     required: true
>                     valid_values:
>                       - pipe
>               php_model:
>                 description: Used on PHP router in the absence of any VPN routes and explicit null VRF labels.
>                 type: dict
>                 keys:
>                   ttl:
>                     type: str
>                     required: true
>                     valid_values:
>                       - pipe
>                       - uniform
>                   dscp:
>                     type: str
>                     required: true
>                     valid_values:
>                       - pipe
>                       - uniform
> ```



> > > ```
> > > tunnel:
> > >   termination:
> > >     - php: < true | false >
> > >       model:   
> > >         ttl: < pipe | uniform >
> > >         dscp: < pipe | uniform >
> > > ```
> > 
> > 
> > Or something explicit like this:
> > ```yaml
    tunnel:
        description: Configure MPLS tunnel.
        type: dict
        keys:
          termination:
            description: Controls selection of the TTL/DSCP values by LER when decapsulating MPLS packets.
            type: dict
            keys:
              model:
                type: dict
                keys:
                  ttl:
                    type: str
                    required: true
                    valid_values:
                      - pipe
                      - uniform
                  dscp:
                    description: DSCP model `uniform` supported with ttl model `uniform` on some platforms.
                    type: str
                    required: true
                    valid_values:
                      - pipe
                      - uniform
              php_model:
                description: Used on PHP router in the absence of any VPN routes and explicit null VRF labels.
                type: dict
                keys:
                  ttl:
                    type: str
                    required: true
                    valid_values:
                      - pipe
                      - uniform
                  dscp:
                    type: str
                    required: true
                    valid_values:
                      - pipe
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I was not able to identify any specific platforms which support `dscp uniform` when configuring `mpls tunnel termination model ttl`. I'll try to figure this out . If not - we may need to only permit `pipe` under `mpls.tunnel.termination.model.dscp`.
> 
> Hi Alexey,
> 
> The explicit version also looks fine to me.
> 
> Regarding compatibility. Check this:
> 
> ```
> X(config-s-test)#mpls tunnel termination model ttl pipe dscp ?
>   pipe     Preserve the inner value
>   uniform  Propagate value from outer header
> 
> X(config-s-test)#mpls tunnel termination model ttl pipe dscp uniform
> % ttl model pipe is not supported with dscp model uniform
> 
> X(config-s-test)#sh version
> Arista DCS-7280CR2A-30-F
> Hardware version: 21.00
> <snip>
> Software image version: 4.26.3M
> ```
> 
> On 4.26.3 the "dscp uniform" was suggested by eos, but couldn't be configured (not supported)
> 
> When i check on a newer EOS:
> 
> ```
> X(s1)(config-s-test)#mpls tunnel termination model ttl pipe dscp ?
>   pipe  Preserve the inner value
> 
> X(s1)(config-s-test)#sh ver
> Arista 7289-F
> <snip>
> Software image version: 4.31.4M
> ```
> 
> not suggesting dscp > uniform anymore.

Hi @alexeygorbunov and @ccsnw 
Added minor changes. 
1. Added uniform as valid values in model dscp as command `mpls tunnel termination model ttl uniform dscp uniform` supported on some platforms.
2. Removed uniform as valid value from php dscp model. 
```
    tunnel:
        description: Configure MPLS tunnel.
        type: dict
        keys:
          termination:
            description: Controls selection of the TTL/DSCP values by LER when decapsulating MPLS packets.
            type: dict
            keys:
              model:
                type: dict
                keys:
                  ttl:
                    type: str
                    required: true
                    valid_values:
                      - pipe
                      - uniform
                  dscp:
                    description: DSCP model `uniform` supported with ttl model `uniform` on some platforms.
                    type: str
                    required: true
                    valid_values:
                      - pipe
                      - uniform
              php_model:
                description: Used on PHP router in the absence of any VPN routes and explicit null VRF labels.
                type: dict
                keys:
                  ttl:
                    type: str
                    required: true
                    valid_values:
                      - pipe
                      - uniform
                  dscp:
                    type: str
                    required: true
                    valid_values:
                      - pipe
```

ccsnw commented:
> > ```yaml
> > tunnel:
> >         description: Configure MPLS tunnel.
> >         type: dict
> >         keys:
> >           termination:
> >             description: Controls selection of the TTL/DSCP values by LER when decapsulating MPLS packets.
> >             type: dict
> >             keys:
> >               model:
> >                 type: dict
> >                 keys:
> >                   ttl:
> >                     type: str
> >                     required: true
> >                     valid_values:
> >                       - pipe
> >                       - uniform
> >                   dscp:
> >                     type: str
> >                     required: true
> >                     valid_values:
> >                       - pipe
> >               php_model:
> >                 description: Used on PHP router in the absence of any VPN routes and explicit null VRF labels.
> >                 type: dict
> >                 keys:
> >                   ttl:
> >                     type: str
> >                     required: true
> >                     valid_values:
> >                       - pipe
> >                       - uniform
> >                   dscp:
> >                     type: str
> >                     required: true
> >                     valid_values:
> >                       - pipe
> >                       - uniform
> > ```
> 
> > > > ```
> > > > tunnel:
> > > >   termination:
> > > >     - php: < true | false >
> > > >       model:   
> > > >         ttl: < pipe | uniform >
> > > >         dscp: < pipe | uniform >
> > > > ```
> > > 
> > > 
> > > Or something explicit like this:
> 
> ```
> tunnel:
>     description: Configure MPLS tunnel.
>     type: dict
>     keys:
>       termination:
>         description: Controls selection of the TTL/DSCP values by LER when decapsulating MPLS packets.
>         type: dict
>         keys:
>           model:
>             type: dict
>             keys:
>               ttl:
>                 type: str
>                 required: true
>                 valid_values:
>                   - pipe
>                   - uniform
>               dscp:
>                 description: DSCP model `uniform` supported with ttl model `uniform` on some platforms.
>                 type: str
>                 required: true
>                 valid_values:
>                   - pipe
>                   - uniform
>           php_model:
>             description: Used on PHP router in the absence of any VPN routes and explicit null VRF labels.
>             type: dict
>             keys:
>               ttl:
>                 type: str
>                 required: true
>                 valid_values:
>                   - pipe
>                   - uniform
>               dscp:
>                 type: str
>                 required: true
>                 valid_values:
>                   - pipe
> ```
> 
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > I was not able to identify any specific platforms which support `dscp uniform` when configuring `mpls tunnel termination model ttl`. I'll try to figure this out . If not - we may need to only permit `pipe` under `mpls.tunnel.termination.model.dscp`.
> > > ```
> > 
> > 
> > Hi Alexey,
> > The explicit version also looks fine to me.
> > Regarding compatibility. Check this:
> > ```
> > X(config-s-test)#mpls tunnel termination model ttl pipe dscp ?
> >   pipe     Preserve the inner value
> >   uniform  Propagate value from outer header
> > 
> > X(config-s-test)#mpls tunnel termination model ttl pipe dscp uniform
> > % ttl model pipe is not supported with dscp model uniform
> > 
> > X(config-s-test)#sh version
> > Arista DCS-7280CR2A-30-F
> > Hardware version: 21.00
> > <snip>
> > Software image version: 4.26.3M
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > On 4.26.3 the "dscp uniform" was suggested by eos, but couldn't be configured (not supported)
> > When i check on a newer EOS:
> > ```
> > X(s1)(config-s-test)#mpls tunnel termination model ttl pipe dscp ?
> >   pipe  Preserve the inner value
> > 
> > X(s1)(config-s-test)#sh ver
> > Arista 7289-F
> > <snip>
> > Software image version: 4.31.4M
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > not suggesting dscp > uniform anymore.
> 
> Hi @alexeygorbunov and @ccsnw Added minor changes.
> 
> 1. Added uniform as valid values in model dscp as command `mpls tunnel termination model ttl uniform dscp uniform` supported on some platforms.
> 2. Removed uniform as valid value from php dscp model.
> 
> ```
>     tunnel:
>         description: Configure MPLS tunnel.
>         type: dict
>         keys:
>           termination:
>             description: Controls selection of the TTL/DSCP values by LER when decapsulating MPLS packets.
>             type: dict
>             keys:
>               model:
>                 type: dict
>                 keys:
>                   ttl:
>                     type: str
>                     required: true
>                     valid_values:
>                       - pipe
>                       - uniform
>                   dscp:
>                     description: DSCP model `uniform` supported with ttl model `uniform` on some platforms.
>                     type: str
>                     required: true
>                     valid_values:
>                       - pipe
>                       - uniform
>               php_model:
>                 description: Used on PHP router in the absence of any VPN routes and explicit null VRF labels.
>                 type: dict
>                 keys:
>                   ttl:
>                     type: str
>                     required: true
>                     valid_values:
>                       - pipe
>                       - uniform
>                   dscp:
>                     type: str
>                     required: true
>                     valid_values:
>                       - pipe
> ```

Sure! Is this final? Then I can redo the rest accordingly. 

Thanks!!

ccsnw committed:
add_mpls_tunnel_termination

ccsnw committed:
add_mpls_tunnel_termination

ccsnw commented:
@alexeygorbunov & @MaheshGSLAB 

updated as discussed :-)

let me know if this looks ok or if more changes are needed!

MaheshGSLAB commented on python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/mpls.schema.yml:
```diff
@@ -282,3 +282,42 @@ keys:
+                    type: str
+                    required: true
+                    valid_values:
+                      - pipe
```
```suggestion
                      - pipe
                      - uniform
```

MaheshGSLAB commented on python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/mpls.schema.yml:
```diff
@@ -282,3 +282,42 @@ keys:
+                      - pipe
+                      - uniform
+                  dscp:
+                    type: str
```
```suggestion
                    description: The DSCP model `uniform` is supported only on specific hardware platforms.
                    type: str
```

MaheshGSLAB commented on python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/mpls.schema.yml:
```diff
@@ -282,3 +282,42 @@ keys:
+                      - pipe
+                      - uniform
+                  dscp:
+                    description: DSCP model `uniform` supported with ttl model `uniform` on some platforms.
```
```suggestion
                    description: The DSCP model `uniform` is supported only on specific hardware platforms.
```

MaheshGSLAB commented:
> @alexeygorbunov & @MaheshGSLAB
> 
> updated as discussed :-)
> 
> let me know if this looks ok or if more changes are needed!

Since the DSCP `uniform` supported combinations depends on the hardware platform i requested some changes in description and valid values

ccsnw committed:
add_mpls_tunnel_termination

ClausHolbechArista committed:
Merge branch 'devel' into add_mpls_tunnel_termination

ClausHolbechArista commented:
LGTM. Thank you for another great contribution!
Sorry for the delay in getting this reviewed. At least it will make it into the 5.2 release :)
